File: find_k_closest_elements_two_pointers.py
# Time Complexity : O(n - k) + O(k) where n length of the array and k is number closest elements to findout
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:
        low = 0
        high = len(arr) - 1
        
        while high - low + 1 > k:
            if x - arr[low] < arr[high] - x:
                high -= 1
            elif x - arr[low] > arr[high] - x:
                low += 1
            else:
                high -= 1
                
            
        return arr[low: high + 1]

File: findKClosestElements_TwoPointers.cpp
// Time Complexity : O(n-k)

// Space Complexity : O(1)

// Did this code successfully run on Leetcode : YES

// Appoarch: Using two pointers to check the dist diff and keep moving pointers until we reach k.

// 658. Find K Closest Elements

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int n = arr.size();
        vector<int> result;
        int left = 0;
        int right = n-1;
        while(right - left >= k){
            int distR = abs(arr[right] - x);
            int distL = abs(arr[left] - x);
            if(distR < distL){
                left++;
            }
            else{
                right--;
            }
        }
        for(int i=left;i<=right;i++){
            result.push_back(arr[i]);
        }
        return result;
    }
};